### ISSUE TITLE
您好，关于在SEED-VIG上进行finetune的疑问（Hello, I have some questions about finetuning on SEED-VIG.）

### ISSUE THREAD

**Issue by INGg** [935963004/LaBraM#56]
在验证集和测试集上的准确率始终不变，请问这是为什么，结果如下（The accuracy on the validation set and test set remains unchanged. Could you please tell me why? The results are as follows.）：

<img width="840" alt="image" src="https://github.com/user-attachments/assets/998274f2-d821-4b22-89cf-f005b1280c18" />

<img width="1265" alt="image" src="https://github.com/user-attachments/assets/36c8dab5-83f2-4541-99df-a6ba76f2ac1e" />

<img width="1251" alt="image" src="https://github.com/user-attachments/assets/303b2bd9-d19a-4511-9f07-f1c2d00a695b" />

<img width="1261" alt="image" src="https://github.com/user-attachments/assets/8155be01-c617-4224-bfa6-ea93d054f157" />

如图所见，loss是在下降的，但是准确率一直是瞎猜的水平，这可能是什么原因呢？有没有其他朋友在SEED-VIG上试过结果呢？
（As can be seen from the figure, the loss is decreasing, but the accuracy has always been at the level of random guessing. What could be the possible reasons for this? Have any other friends tried it on SEED-VIG and obtained results?）
***

处理代码如下：
（The code for processing is as follows:）
```
def prepare_SEEDVIG_dataset(root):
    # set random seed
    seed = 12345
    np.random.seed(seed)


    sub_list = os.listdir(os.path.join(root, "Raw_Data"))
    for sub in sub_list:
        # reading mat
        mat = scipy.io.loadmat(os.path.join(root, "Raw_Data", sub))["EEG"]["data"][0][0].T  # (17, 1416000)


        mat = mat / 1e6
        
        # create raw
        fs = 200
        ch_names = ['FT7', 'FT8', 'T7', 'T8', 'TP7', 'TP8', 'CP1', 'CP2', 'P1', 'PZ', 'P2', 'PO3', 
                        'POZ', 'PO4', 'O1', 'OZ', 'O2']
        info = mne.create_info(ch_names=ch_names, sfreq=fs, ch_types='eeg')
        raw = mne.io.RawArray(mat, info)

        # filtering
        raw = raw.filter(l_freq=0.1, h_freq=75)
        raw = raw.notch_filter(50.0)
        # downsampling
        # raw = raw.resample(200, n_jobs=5)
        eegData = raw.get_data(units='uV')

        temp = np.zeros((17, 885, 1600))
        for i in range(885):
            temp[:, i, :] = eegData[:, i*1600: (i+1)*1600]
        eegData = temp

        # z-score
        mean = np.mean(eegData)
        std = np.std(eegData)
        eegData = (eegData - mean) / std

        # label
        perclos = scipy.io.loadmat(os.path.join(root, "perclos_labels", sub))["perclos"]   # (885, 1)
        temp = np.zeros([885, 1], dtype=int)
        for i in range(885):
            if perclos[i] < 0.35:
                temp[i] = 0  # awake
            elif 0.35 <= perclos[i] < 0.7:
                temp[i] = 1  # tired
            else:
                temp[i] = 2  # drowsy
        # label = np.eye(3)[temp]                                                     # (885, 1, 3), one-hot encoding
        label = temp

        print(eegData.shape, label.shape)

        # 按照60%，20%，20%划分训练、验证、测试集
        index = np.random.permutation(label.shape[0])
        # print(0, int(0.6*label.shape[0]), int(0.8*label.shape[0]), label.shape[0])
        train_index = index[0: int(0.6*label.shape[0])]
        eval_index = index[int(0.6*label.shape[0]): int(0.8*label.shape[0])]
        test_index = index[int(0.8*label.shape[0]): label.shape[0]]

        train_eeg = eegData[:, train_index, :]
        train_label = label[train_index]
        eval_eeg = eegData[:, eval_index, :]
        eval_label = label[eval_index]
        test_eeg = eegData[:, test_index, :]
        test_label = label[test_index]

        train_eeg = torch.FloatTensor(train_eeg).permute(1, 0, 2)
        # train_eeg = torch.flatten(train_eeg, start_dim=1)
        train_label = torch.FloatTensor(train_label)
        train_dataset = TensorDataset(train_eeg, train_label)

        eval_eeg = torch.FloatTensor(eval_eeg).permute(1, 0, 2)
        # eval_eeg = torch.flatten(eval_eeg, start_dim=1)
        eval_label = torch.FloatTensor(eval_label)
        eval_dataset = TensorDataset(eval_eeg, eval_label)

        test_eeg = torch.FloatTensor(test_eeg).permute(1, 0, 2)
        # test_eeg = torch.flatten(test_eeg, start_dim=1)
        test_label = torch.FloatTensor(test_label)
        test_dataset = TensorDataset(test_eeg, test_label)
        
        return train_dataset, test_dataset, eval_dataset
        
    print(sub_list)
    # train_files = os.listdir(os.path.join(root, "train"))
    # np.random.shuffle(train_files)
    # val_files = os.listdir(os.path.join(root, "val"))
    # test_files = os.listdir(os.path.join(root, "test"))

    # print(len(train_files), len(val_files), len(test_files))

    # # prepare training and test data loader
    # train_dataset = TUABLoader(os.path.join(root, "train"), train_files)
    # test_dataset = TUABLoader(os.path.join(root, "test"), test_files)
    # val_dataset = TUABLoader(os.path.join(root, "val"), val_files)
    # print(len(train_files), len(val_files), len(test_files))
    # return train_dataset, test_dataset, val_dataset
```

`run_class_finetune`中的修改如下：
（The modifications in run_class_finetune are as follows:）
```
def get_dataset(args):
    if args.dataset == 'TUAB':
        train_dataset, test_dataset, val_dataset = utils.prepare_TUAB_dataset("path/to/TUAB")
        ch_names = ['EEG FP1', 'EEG FP2-REF', 'EEG F3-REF', 'EEG F4-REF', 'EEG C3-REF', 'EEG C4-REF', 'EEG P3-REF', 'EEG P4-REF', 'EEG O1-REF', 'EEG O2-REF', 'EEG F7-REF', \
                    'EEG F8-REF', 'EEG T3-REF', 'EEG T4-REF', 'EEG T5-REF', 'EEG T6-REF', 'EEG A1-REF', 'EEG A2-REF', 'EEG FZ-REF', 'EEG CZ-REF', 'EEG PZ-REF', 'EEG T1-REF', 'EEG T2-REF']
        ch_names = [name.split(' ')[-1].split('-')[0] for name in ch_names]
        args.nb_classes = 1
        metrics = ["pr_auc", "roc_auc", "accuracy", "balanced_accuracy"]
    
    elif args.dataset == 'TUEV':
        train_dataset, test_dataset, val_dataset = utils.prepare_TUEV_dataset("path/to/TUEV")
        ch_names = ['EEG FP1-REF', 'EEG FP2-REF', 'EEG F3-REF', 'EEG F4-REF', 'EEG C3-REF', 'EEG C4-REF', 'EEG P3-REF', 'EEG P4-REF', 'EEG O1-REF', 'EEG O2-REF', 'EEG F7-REF', \
                    'EEG F8-REF', 'EEG T3-REF', 'EEG T4-REF', 'EEG T5-REF', 'EEG T6-REF', 'EEG A1-REF', 'EEG A2-REF', 'EEG FZ-REF', 'EEG CZ-REF', 'EEG PZ-REF', 'EEG T1-REF', 'EEG T2-REF']
        ch_names = [name.split(' ')[-1].split('-')[0] for name in ch_names]
        args.nb_classes = 6
        metrics = ["accuracy", "balanced_accuracy", "cohen_kappa", "f1_weighted"]

    elif args.dataset == 'SEED-VIG':
        train_dataset, test_dataset, val_dataset = utils.prepare_SEEDVIG_dataset("/work/lizhengde/code/dataset/SEED-VIG")
        ch_names = ['FT7', 'FT8', 'T7', 'T8', 'TP7', 'TP8', 'CP1', 'CP2', 'P1', 'PZ', 'P2', 'PO3', 
                    'POZ', 'PO4', 'O1', 'OZ', 'O2']
        args.nb_classes = 3
        metrics = ["accuracy", "balanced_accuracy", "cohen_kappa", "f1_weighted"]
    return train_dataset, test_dataset, val_dataset, ch_names, metrics
```

运行参数为：
（The running parameters are:）
```
OMP_NUM_THREADS=1 torchrun --nnodes=1 --nproc_per_node=2 run_class_finetuning.py \
        --output_dir ./checkpoints/finetune_tuab_base/ \
        --log_dir ./log/finetune_tuab_base \
        --model labram_base_patch200_200 \
        --finetune ./checkpoints/labram-base.pth \
        --weight_decay 0.05 \
        --batch_size 64 \
        --lr 5e-4 \
        --update_freq 1 \
        --warmup_epochs 3 \
        --epochs 30 \
        --layer_decay 0.65 \
        --drop_path 0.1 \
        --dist_eval \
        --save_ckpt_freq 5 \
        --disable_rel_pos_bias \
        --abs_pos_emb \
        --dataset SEED-VIG \
        --disable_qkv_bias \
        --seed 0
  ```

**Comment by yuran986** [comment]
遇到了类似问题，甚至我这里val_loss也不变，train_loss则是一直振荡，求问解决了吗

**Comment by INGg** [comment]
> 遇到了类似问题，甚至我这里val_loss也不变，train_loss则是一直振荡，求问解决了吗

并没有，但是有的文献用来比较的结果复现的还可以
